src/compiler/crystal/codegen/fun.cr

```diff
@@ -382,7 +395,14 @@ class Crystal::CodeGenVisitor
   def setup_context_fun(mangled_name, target_def, llvm_args_types, llvm_return_type) : Nil
     context.fun = @llvm_mod.functions.add(mangled_name, llvm_args_types, llvm_return_type, target_def.varargs?)
 
-    context.fun.add_attribute LLVM::Attribute::AlwaysInline if target_def.always_inline?
+    unless @debug.variables?
+      context.fun.add_attribute LLVM::Attribute::AlwaysInline if target_def.always_inline?
+    else
+      context.fun.add_attribute LLVM::Attribute::NoInline
+      context.fun.add_attribute LLVM::Attribute::NoUnwind
+      context.fun.add_attribute LLVM::Attribute::OptimizeNone
+      context.fun.add_attribute LLVM::Attribute::UWTable
```

Comment:
Can we really not avoid these de-optimizations when passing `--debug`?

Reply:
I copied them from clang debug IR code. For debug versions they are necessary to show correctly debug values. In debug mode you are not expecting performance but correctness of debug values.

Reply:
`--debug` is variables, default is line numbers, `--no-debug` is nothing.

I think it's fine that `--debug` has a performance penalty, though I wonder what th rest of the core team thinks.

Reply:
I think the `NoUnwind` and `UWTable` attributes are incorrect here, but the `NoInline` and `OptimizeNone` are fine. Are you sure all of these are required.

Have you tested raising in `--debug` code?

Reply:
Here is the LLVM IR with debug info for clang compiled llvm_ext.o.ll:
```
; Function Attrs: noinline nounwind optnone ssp uwtable
define internal i8* @_ZNSt3__1L9addressofIKcEEPT_RS2_(i8* dereferenceable(1)) #0 !dbg !77195 {
  %2 = alloca i8*, align 8
  store i8* %0, i8** %2, align 8
  call void @llvm.dbg.declare(metadata i8** %2, metadata !77198, metadata !DIExpression()), !dbg !77199
  %3 = load i8*, i8** %2, align 8, !dbg !77200
  ret i8* %3, !dbg !77201
}

attributes #0 = { noinline nounwind optnone ssp uwtable "correctly-rounded-divide-sqrt-fp-math"="false" "disable-tail-calls"="false" "less-precise-fpmad"="false" "min-legal-vector-width"="0" "no-frame-poi
nter-elim"="true" "no-frame-pointer-elim-non-leaf" "no-infs-fp-math"="false" "no-jump-tables"="false" "no-nans-fp-math"="false" "no-signed-zeros-fp-math"="false" "no-trapping-math"="false" "stack-protecto
r-buffer-size"="8" "target-cpu"="penryn" "target-features"="+cx16,+fxsr,+mmx,+sahf,+sse,+sse2,+sse3,+sse4.1,+ssse3,+x87" "unsafe-fp-math"="false" "use-soft-float"="false" }
```
I just followed this way. I added not all of them but only the ones I think relevant in this case. 

Exceptions are rising just fine as you can see in this test response:
```
Unhandled exception: Error is here (Exception)
  from debug_tests.cr:24:3 in '__crystal_main'
  from /Users/sergey/Projects/crystal/crystal/src/crystal/main.cr:106:5 in 'main_user_code'
  from /Users/sergey/Projects/crystal/crystal/src/crystal/main.cr:92:7 in 'main'
  from /Users/sergey/Projects/crystal/crystal/src/crystal/main.cr:115:3 in 'main'
```

Reply:
If you compile

```cpp
int square(int num) {
    throw "foo";
}
```

then clang doesn't add `nounwind`.

Neither do the "uwtable" or "nounwind" attributes change between `-O0` and `-O3`, so they are unrelated to debug support and should be removed.

Reply:
Good catch! OK, I will remove them.

Reply:
Here's the part of clang where it adds these attributes: https://github.com/llvm/llvm-project/blob/fc3752665f4b7ae2ec3faad4204512687800c1ec/clang/lib/CodeGen/CodeGenModule.cpp#L1586

the logic is, if we're in `-O0` (`ShouldAddOptNone`) and the function *isn't* marked `AlwaysInline`, then add `OptimizeNone` and `NoInline`. Other function attributes are unrelated.

Reply:
Sorry for these last minute changes again!

Reply:
No problem at all! The more stable the better.
Did you played with this branch yourself? Do you have any concerns or suggestions?
It not supporting global and class variables yet though.

Reply:
I have played with the branch for https://github.com/crystal-lang/crystal/pull/4538, which is waiting on this PR. It worked great!

I think incomplete debug support is fine, this PR is a great improvement, I just want to avoid any other possible regressions.

Reply:
I think your tool is invaluable as we can implement profile based branch optimization.